File: system/tests/rspec/pages/files/upload_edit.rb
class UploadEdit < ControlPanelPage

	element :name, 'input[name=name]'
	element :url, 'input[name=url]'
	element :server_path, 'input[name=server_path]'
	element :allowed_types, 'select[name=allowed_types]'
	element :max_size, 'input[name=max_size]'
	element :max_width, 'input[name=max_width]'
	element :max_height, 'input[name=max_height]'
	element :image_manipulations, '#image_manipulations'
	elements :grid_rows, '#image_manipulations tr'
	elements :upload_member_groups, 'input[name="upload_member_groups[]"]'
	elements :cat_group, 'input[name="cat_group[]"]'

	def load
		click_link 'Files'
		within 'div.sidebar h2:nth-child(1)' do
			click_link 'New'
		end
	end

	def load_edit_for_dir(number)
		click_link 'Files'

		directory_selector = 'div.sidebar .folder-list > li:nth-child('+number.to_s+')'

		find(directory_selector).hover
		find(directory_selector+' li.edit a').click
	end

	# Dynamic getter for a specific Grid row
	def grid_row(row)
		# Plus three to skip over header, blank row and no results row
		image_manipulations.find('tbody tr:nth-child('+(row+2).to_s+')')
	end

	# Returns the name field in a specific Grid row, and so on...
	def name_for_row(row)
		self.grid_row(row).find('td:first-child input')
	end

	def resize_type_for_row(row)
		self.grid_row(row).find('td:nth-child(2) select')
	end

	def width_for_row(row)
		self.grid_row(row).find('td:nth-child(3) input')
	end

	def height_for_row(row)
		self.grid_row(row).find('td:nth-child(4) input')
	end

	def delete_for_row(row)
		self.grid_row(row).find('td:nth-child(5) li.remove a')
	end
end

File: tests/cypress/cypress/elements/pages/files/UploadEdit.js
import FileManagerSection from '../_sections/FileManagerSection'

class UploadEdit extends FileManagerSection {
  constructor() {
      super()
      this.urlMatch = /files\/file\/edit/;

      this.elements({
        'name': 'input[type!=hidden][name=name]',
        'url': 'input[type!=hidden][name="_for_adapter[local][url]"]',
        'adapter': 'div[data-input-value=adapter]',
        'server_path': 'input[type!=hidden][name="_for_adapter[local][server_path]"]',
        'allowed_types': 'div[data-input-value="allowed_types"]',
        'max_size': 'input[type!=hidden][name=max_size]',
        'max_width': 'input[type!=hidden][name=max_width]',
        'max_height': 'input[type!=hidden][name=max_height]',
        'image_manipulations': '#image_manipulations',
        'grid_rows': '#image_manipulations tr:visible',
        'upload_member_groups': 'input[type!=hidden][name="upload_roles[]"]',
        'cat_group': 'input[type!=hidden][name="cat_group[]"]'
      })
    }
    load() {
      cy.contains('Files').click()
      cy.get('div.sidebar h2').contains('New').click()
    }

    load_edit_for_dir(number) {
      cy.contains('Files').click()
      cy.get('div.sidebar .folder-list > div:nth-child('+number.toString()+')').trigger('mouseover')
      cy.get('div.sidebar .folder-list > div:nth-child('+number.toString()+')  a.edit').click({force: true})
    }

    create_manipulation() {

      this.load_edit_for_dir(2)

      this.get('grid_add_no_results').click()
      this.name_for_row(1).type('some_name')
      this.width_for_row(1).type('20')
      this.height_for_row(1).type('30')

      this.get('grid_add').click()
      this.name_for_row(2).type('some_other_name')
      this.resize_type_for_row(2).select('Crop (part of image)')
      this.width_for_row(2).type('50')
      this.height_for_row(2).type('40')

      cy.hasNoErrors()

      this.submit()
      this.get('wrap').contains('Upload directory saved')

      cy.hasNoErrors()
    }

    // Dynamic getter for a specific Grid row
    grid_row(row) {
      // Plus three to skip over header, blank row and no results row
      return this.get('image_manipulations').find('tbody tr:nth-child('+(row+2).toString()+')')
    }

    // Returns the name field in a specific Grid row, and so on...
    name_for_row(row) {
      return this.grid_row(row).find('td:first-child input')
    }

    resize_type_for_row(row) {
      return this.grid_row(row).find('td:nth-child(2) select')
    }

    width_for_row(row) {
      return this.grid_row(row).find('td:nth-child(3) input')
    }

    height_for_row(row) {
      return this.grid_row(row).find('td:nth-child(4) input')
    }

    quality_for_row(row) {
      return this.grid_row(row).find('td:nth-child(5) input')
    }

    watermark_for_row(row) {
      return this.grid_row(row).find('td:nth-child(6) select')
    }

    delete_for_row(row) {
      return this.grid_row(row).find('td:nth-child(7) [rel=remove_row]')
    }
}
export default UploadEdit;
